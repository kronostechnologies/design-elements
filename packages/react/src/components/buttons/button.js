"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var styled_components_1 = __importDefault(require("styled-components"));
var primary_1 = require("./styles/primary");
var secondary_1 = require("./styles/secondary");
var tertiary_1 = require("./styles/tertiary");
var abstract_button_1 = require("./abstract-button");
var StyledButton = styled_components_1.default(abstract_button_1.AbstractButton)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    ", "\n"], ["\n    ",
    "\n"])), function (props) {
    if (props.buttonType === 'secondary') {
        return secondary_1.secondaryStyles;
    }
    else if (props.buttonType === 'tertiary') {
        return tertiary_1.tertiaryStyles;
    }
    return primary_1.primaryStyles;
});
exports.Button = function (_a) {
    var children = _a.children, disabled = _a.disabled, onClick = _a.onClick, buttonType = _a.buttonType;
    var handleClick = function () { onClick && onClick(); };
    return (<StyledButton disabled={disabled} onClick={handleClick} buttonType={buttonType}>
            {children}
        </StyledButton>);
};
var templateObject_1;
