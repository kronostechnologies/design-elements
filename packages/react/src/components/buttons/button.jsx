import React from 'react';
import styled from 'styled-components';

import { SECONDARY, TERTIARY } from '../../constants';

import primaryStyle from './styles/primary';
import secondaryStyle from './styles/secondary';
import tertiaryStyle from './styles/tertiary';

import AbstractButton from './abstract-button';

const StyledButton = styled(AbstractButton)`
  ${props => {
        if (props.type === SECONDARY) {
            return secondaryStyle;
        } else if (props.type === TERTIARY) {
            return tertiaryStyle;
        }
        return primaryStyle;
    }}
`;

const Button = ({ children, disabled, onClick }) => {
    const handleClick = () => {
        if (typeof onClick === 'function') {
            onClick();
        }
    };

    return (
        <StyledButton disabled={disabled} onClick={handleClick}>
            { children }
        </StyledButton>
    );
};

export default Button;
