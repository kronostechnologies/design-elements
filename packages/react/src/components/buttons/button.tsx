import React, { ReactElement } from 'react';

import styled from 'styled-components';
import { Theme } from '../theme-wrapper/theme-wrapper';
import { AbstractButton, AbstractButtonProps } from './abstract-button';

type ButtonType = 'primary' | 'secondary' | 'tertiary';

type Type = 'submit' | 'button' | 'reset';

interface ButtonProps extends AbstractButtonProps {
    /**
     * Visual style
     * @default primary
     **/
    buttonType: ButtonType;
    /**
     * Sets button type
     * @default submit
     */
    type?: Type;
}

const StyledButton = styled(AbstractButton)<ButtonProps>`
    ${(props: { theme: Theme, buttonType: ButtonType }) => {
        switch (props.buttonType) {
            case 'primary':
                return `
                    background-color: ${props.theme.main['primary-1.1']};
                    border-color: ${props.theme.main['primary-1.1']};
                    color: ${props.theme.greys.white};

                    &:hover {
                        background-color: ${props.theme.main['primary-1.3']};
                        border-color: ${props.theme.main['primary-1.3']};
                    }
                    &:focus {
                        box-shadow: 0 0 0 2px rgba(0, 128, 165, 0.4);
                    }

                    &:disabled {
                        &,
                        &:focus,
                        &:hover {
                            background-color: ${props.theme.main['primary-1.2']};
                            border-color: ${props.theme.main['primary-1.2']};
                        }
                    }
                  `;
            case 'secondary':
                return `
                    background-color: transparent;
                    border-color: ${props.theme.main['primary-1.1']};
                    color: ${props.theme.main['primary-1.1']};

                    &:hover {
                        border-color: ${props.theme.main['primary-1.3']};
                        color: ${props.theme.main['primary-1.3']};
                    }
                    &:focus {
                        box-shadow: 0 0 0 2px rgba(0, 128, 165, 0.4);
                    }

                    &:disabled {
                        &,
                        &:focus,
                        &:hover {
                            border-color: ${props.theme.main['primary-1.2']};
                            color: ${props.theme.main['primary-1.2']};
                        }
                    }
                `;
            case 'tertiary':
                return `
                    background-color: transparent;
                    border-color: transparent;
                    color: ${props.theme.greys['dark-grey']};

                    &:hover {
                        background-color: ${props.theme.greys.grey};
                        color: ${props.theme.greys.black};
                    }
                    &:focus {
                        border-color: ${props.theme.main['primary-1.1']};
                        box-shadow: 0 0 0 2px rgba(0, 128, 165, 0.4);
                    }

                    &:disabled {
                        &,
                        &:focus,
                        &:hover {
                            background-color: transparent;
                            color: ${props.theme.greys['mid-grey']};
                        }
                    }
                `;
        }
    }
}
`;

export function Button({ children, onClick, type = 'submit', ...props }: ButtonProps): ReactElement {
    function handleClick(): void {
        onClick && onClick();
    }

    const label: string = props.label || '';
    return (
        <StyledButton onClick={handleClick} type={type} {...props}>
            {children}{label}
        </StyledButton>
    );
}
