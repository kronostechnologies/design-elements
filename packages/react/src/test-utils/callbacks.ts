export function doNothing(): void {
    // do nothing
}
