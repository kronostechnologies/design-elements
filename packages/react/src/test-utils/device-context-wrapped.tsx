import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
