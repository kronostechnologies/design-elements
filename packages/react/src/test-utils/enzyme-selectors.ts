import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
