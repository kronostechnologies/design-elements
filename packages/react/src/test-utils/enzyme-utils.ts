import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
