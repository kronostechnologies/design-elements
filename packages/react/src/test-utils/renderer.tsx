import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
