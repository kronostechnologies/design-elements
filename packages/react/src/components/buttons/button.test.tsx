import { doNothing } from '../../test-utils/callbacks';
import { mountWithProviders, renderWithProviders } from '../../test-utils/renderer';
import { Button } from './button';

describe('Button', () => {
    test('onClick callback is called when clicked', () => {
        const callback = jest.fn();
        const wrapper = mountWithProviders(<Button onClick={callback} buttonType="primary" label="Primary Button" />);

        wrapper.simulate('click');

        expect(callback).toHaveBeenCalledTimes(1);
    });

    test('onClick callback cannot be called when disabled', () => {
        const callback = jest.fn();
        const wrapper = mountWithProviders(
            <Button
                onClick={callback}
                buttonType="primary"
                disabled
                label="Primary Button"
            />,
        );

        wrapper.simulate('click');

        expect(callback).toHaveBeenCalledTimes(0);
    });

    test('has primary disabled styles', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="primary" disabled label="Primary Button" />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has primary styles', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="primary" label="Primary Button" />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has primary styles (inverted)', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="primary" label="Primary Button" inverted />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has secondary styles', () => {
        const tree = renderWithProviders(
            <Button
                onClick={doNothing}
                buttonType="secondary"
                label="Secondary Button"
            />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has secondary styles (inverted)', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="secondary" label="Secondary Button" inverted />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has tertiary styles', () => {
        const tree = renderWithProviders(<Button onClick={doNothing} buttonType="tertiary" label="Tertiary Button" />);

        expect(tree).toMatchSnapshot();
    });

    test('has tertiary styles (inverted)', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="tertiary" label="Tertiary Button" inverted />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has destructive styles', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="destructive-primary" label="Destructive Button" />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has destructive styles (inverted)', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="destructive-primary" label="Destructive Button" inverted />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has destructive-secondary styles', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="destructive-primary" label="Destructive Button" />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has destructive-secondary styles (inverted)', () => {
        const tree = renderWithProviders(
            <Button
                onClick={doNothing}
                buttonType="destructive-secondary"
                label="Destructive Secondary Button"
                inverted
            />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has small styles', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="primary" label="Small Primary Button" size="small" />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('has mobile styles', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="primary" label="Primary Button" />,
            'mobile',
        );

        expect(tree).toMatchSnapshot();
    });

    test('has small styles on mobile', () => {
        const tree = renderWithProviders(
            <Button onClick={doNothing} buttonType="primary" label="Small Primary Button" size="small" />,
            'mobile',
        );

        expect(tree).toMatchSnapshot();
    });

    test('focusable button has no tabIndex prop', () => {
        const wrapper = mountWithProviders(<Button buttonType="primary" label="Primary Button" />);

        expect(wrapper.getDOMNode().getAttribute('tabIndex')).toBeNull();
    });

    test('non-focusable button has tabIndex=-1', () => {
        const wrapper = mountWithProviders(<Button buttonType="primary" label="Primary Button" focusable={false} />);

        expect(wrapper.getDOMNode().getAttribute('tabIndex')).toBe('-1');
    });

    test('focusable button has focus styles', () => {
        const wrapper = mountWithProviders(<Button buttonType="primary" label="Primary Button" />);

        expect(wrapper).toHaveStyleRule('outline', 'none', {
            modifier: ':focus',
        });
    });

    test('non-focusable button does not have focus styles', () => {
        const wrapper = mountWithProviders(<Button buttonType="primary" label="Primary Button" focusable={false} />);

        expect(wrapper).not.toHaveStyleRule('outline', 'none', {
            modifier: ':focus',
        });
    });
});
