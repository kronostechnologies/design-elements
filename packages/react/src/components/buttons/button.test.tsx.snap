// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Has primary styles 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: inherit;
  border: 1px solid;
  border-radius: 1.6rem;
  color: inherit;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: inherit;
  font-size: 0.75rem;
  font-weight: var(--font-bold);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-letter-spacing: 0.55px;
  -moz-letter-spacing: 0.55px;
  -ms-letter-spacing: 0.55px;
  letter-spacing: 0.55px;
  min-height: 2rem;
  min-width: 2rem;
  outline: none;
  padding: var(--spacing-half) var(--spacing-2x);
  text-transform: uppercase;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #EF483E;
  border-color: #EF483E;
  color: #FFFFFF;
}

.c0:not(:disabled) {
  cursor: pointer;
}

.c0 > svg {
  color: inherit;
}

.c0:hover {
  background-color: #D41F14;
  border-color: #D41F14;
}

.c0:disabled,
.c0:disabled:focus,
.c0:disabled:hover {
  background-color: #F9B6B2;
  border-color: #F9B6B2;
}

<button
  className="c0"
  label="Primary Button"
  onClick={[Function]}
  type="submit"
>
  Primary Button
</button>
`;

exports[`Button Has secondary styles 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: inherit;
  border: 1px solid;
  border-radius: 1.6rem;
  color: inherit;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: inherit;
  font-size: 0.75rem;
  font-weight: var(--font-bold);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-letter-spacing: 0.55px;
  -moz-letter-spacing: 0.55px;
  -ms-letter-spacing: 0.55px;
  letter-spacing: 0.55px;
  min-height: 2rem;
  min-width: 2rem;
  outline: none;
  padding: var(--spacing-half) var(--spacing-2x);
  text-transform: uppercase;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border-color: #EF483E;
  color: #EF483E;
}

.c0:not(:disabled) {
  cursor: pointer;
}

.c0 > svg {
  color: inherit;
}

.c0:hover {
  background-color: #EF483E;
  border-color: #EF483E;
  color: #FFFFFF;
}

.c0:disabled,
.c0:disabled:focus,
.c0:disabled:hover {
  background-color: transparent;
  border-color: #F9B6B2;
  color: #F9B6B2;
}

<button
  className="c0"
  label="Secondary Button"
  onClick={[Function]}
  type="submit"
>
  Secondary Button
</button>
`;

exports[`Button Has tertiary styles 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: inherit;
  border: 1px solid;
  border-radius: 1.6rem;
  color: inherit;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: inherit;
  font-size: 0.75rem;
  font-weight: var(--font-bold);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-letter-spacing: 0.55px;
  -moz-letter-spacing: 0.55px;
  -ms-letter-spacing: 0.55px;
  letter-spacing: 0.55px;
  min-height: 2rem;
  min-width: 2rem;
  outline: none;
  padding: var(--spacing-half) var(--spacing-2x);
  text-transform: uppercase;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border-color: transparent;
  color: #57666E;
}

.c0:not(:disabled) {
  cursor: pointer;
}

.c0 > svg {
  color: inherit;
}

.c0:hover {
  color: #000000;
}

.c0:disabled,
.c0:disabled:focus,
.c0:disabled:hover {
  color: #9CA7B4;
}

<button
  className="c0"
  label="Tertiary Button"
  onClick={[Function]}
  type="submit"
>
  Tertiary Button
</button>
`;

exports[`Button Is disabled 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: inherit;
  border: 1px solid;
  border-radius: 1.6rem;
  color: inherit;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: inherit;
  font-size: 0.75rem;
  font-weight: var(--font-bold);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-letter-spacing: 0.55px;
  -moz-letter-spacing: 0.55px;
  -ms-letter-spacing: 0.55px;
  letter-spacing: 0.55px;
  min-height: 2rem;
  min-width: 2rem;
  outline: none;
  padding: var(--spacing-half) var(--spacing-2x);
  text-transform: uppercase;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #EF483E;
  border-color: #EF483E;
  color: #FFFFFF;
}

.c0:not(:disabled) {
  cursor: pointer;
}

.c0 > svg {
  color: inherit;
}

.c0:hover {
  background-color: #D41F14;
  border-color: #D41F14;
}

.c0:disabled,
.c0:disabled:focus,
.c0:disabled:hover {
  background-color: #F9B6B2;
  border-color: #F9B6B2;
}

<button
  className="c0"
  disabled={true}
  label="Primary Button"
  onClick={[Function]}
  type="submit"
>
  Primary Button
</button>
`;
